File: MinVectorProductByPermutation/FSharpVersion/MinScalarVectorProduct.fs
namespace FSharpVersion

type Vector = seq<int>
    
type MinScalarVectorProduct(first: Vector, second: Vector, dimension : int) = 
    
    let _dimention = dimension
    
    let buildNewVector(vector : Vector, i : int, j : int) : Vector =
        seq { for k in 0 .. Seq.length(vector) - 1 do
                if k = i then yield Seq.item(j) vector else
                    if k = j then yield Seq.item(i) vector else 
                        yield Seq.item(k) vector
        }

    let permutate(vector : Vector) : seq<Vector> = 
        seq { for i in 0 .. _dimention - 1 do
                
                for j in i .. 1 do
                    let newVector = buildNewVector(vector, j - 1, j)
                    yield newVector
                
                for k in i .. _dimention - 2 do
                    let newVector = buildNewVector(vector, i, k + 1)
                    yield newVector
            }

    let scalarProduct(first : Vector, second : Vector): int = 
        Seq.map2 (fun x y -> x * y ) first second |> Seq.sum 

    let getMinScalarProduct firstVectors secondVectors  =

        let first = Array.ofSeq (firstVectors)
        let second = Array.ofSeq (secondVectors)
        
        let length = Array.length(first)

        let mutable min = 2147483647

        for i in 0 .. length - 1 do
            for j in 0 .. length - 1 do
                if i <> j then
                    let currentScalar = scalarProduct(first.[i], second.[j]);
                    if currentScalar < min then min <- currentScalar

        min
        
    let answer = permutate(first), permutate(second) |> getMinScalarProduct

File: MinVectorProductByPermutation/CSharpVersion/MinScalarVectorProduct.cs
using System;
using System.Collections.Generic;

namespace CSharpVersion
{
    public class MinScalarVectorProduct
    {
        private readonly int _dimension;

        public MinScalarVectorProduct(Vector firstVector, Vector secondVector, int dimension)
        {
            _dimension = dimension;

            if (firstVector == null)
                throw new ArgumentException("Не задан первые вектор");

            if (secondVector == null)
                throw new ArgumentException("Не задан второй вектор");

            var firstVectorPermutations = Permutate(firstVector);
            var secondVectorPermutations = Permutate(secondVector);
            var answer = GetMinScalarProduct(firstVectorPermutations, secondVectorPermutations);
            Answer = answer;
        }

        private List<Vector> Permutate(Vector vector)
        {
            var result = new List<Vector>();
            for (int i = 0; i < _dimension; i++)
            {
                int j = i;
                while (j > 0)
                {
                    var newVector = vector.Copy();
                    Swap(newVector, j - 1, j);
                    result.Add(newVector);
                    j--;
                }

                int k = i;
                while (k < _dimension - 1)
                {
                    var newVector = vector.Copy();
                    Swap(newVector, i, k + 1);
                    result.Add(newVector);
                    k++;
                }
            }

            return result;
        }

        private void Swap(Vector newVector, int i, int j)
        {
            var tmp = newVector[j];
            newVector[j] = newVector[i];
            newVector[i] = tmp;
        }

        private int GetMinScalarProduct(List<Vector> first, List<Vector> second)
        {
            var min = int.MaxValue;
            for (int i = 0; i < first.Count; i++)
            {
                for (int j = 0; j < second.Count; j++)
                {
                    if (i == j)
                        continue;
                    var currentScalar = ScalarProduct(first[i], second[j]);
                    if (currentScalar < min)
                    {
                        min = currentScalar;
                    }
                }
            }

            return min;
        }

        private int ScalarProduct(Vector vector1, Vector vector2)
        {
            int sum = 0;
            for (int i = 0; i < _dimension; i++)
            {
                sum += vector1[i] * vector2[i];
            }

            return sum;
        }

        public int Answer { get; set; } = 0;
    }
}
